gr-digital/lib/afsk_mod_bf_impl.cc

```diff
@@ -0,0 +1,189 @@
+}
+
+void afsk_mod_bf_impl::nrz(std::vector<int>& bits)
+{
+    for (size_t bit = 0; bit < bits.size(); bit++) {
+        bits[bit] = bits[bit] * 2 - 1;
+    }
+}
+
+
+int afsk_mod_bf_impl::work(int noutput_items,
+                           gr_vector_const_void_star& input_items,
+                           gr_vector_void_star& output_items)
+{
+    auto in = static_cast<const input_type*>(input_items[0]);
+    auto out = static_cast<output_type*>(output_items[0]);
+
+
+    float freq_center = calc_center_freq(freq_zero, freq_one);
+    float freq_delta = calc_delta_freq(freq_zero, freq_one);
+
+    std::vector<int> sliced_bits;
+    // preallocate vector for better performance
+    sliced_bits.reserve(noutput_items);
+
+    if (!bits) {
+        // Input is not bits, but bytes: Iterate over input bytes and slice
+        // them into separate bits. Store them in a vector afterwards.
+
+        switch (endianness) {
+        case GR_MSB_FIRST:
+            for (int byte = 0; byte < noutput_items / 8; byte++) {
+                uint8_t raw_byte = in[byte];
+
+                for (int i = 7; i >= 0; --i) {
+                    int bit = (raw_byte >> i) & 1;
+                    sliced_bits.push_back(bit);
+                }
+            }
+            break;
+        case GR_LSB_FIRST:
+            for (int byte = 0; byte < noutput_items / 8; byte++) {
+                uint8_t raw_byte = in[byte];
+
+                for (int i = 0; i <= 7; ++i) {
+                    int bit = (raw_byte >> i) & 1;
+                    sliced_bits.push_back(bit);
+                }
+            }
+            break;
+        default:
+            this->d_logger->error("unknown endianness");
+            throw std::runtime_error("unknown endianness");
+        }
+    } else {
+        // move input_bits into sliced bits
+        for (int bit = 0; bit < noutput_items; bit++) {
+            sliced_bits.push_back(in[bit]);
+        }
+    }
+
+    /* use implementation from https://notblackmagic.com/bitsnpieces/afsk/
+     * as blueprint
+     */
+
+    // Convert bits to NRZ signal
+    nrz(sliced_bits);
+
+    // each bit will get a certain amount of samples in the output signal
+    int steps = samp_rate / bit_rate;
+
+    // These get used inside the loop. Reside here for avoiding multiple
+    // divisions (increases performance)
+    double fd_sr = freq_delta / samp_rate;
+    double fc_sr = freq_center / samp_rate;
+
+    for (int i = 1; i < noutput_items; i++) {
+        // calculate the source bits indices in bit-vector on basis of
+        // output-buffer-size. These are needed for the integral
+        int curr_i = i / steps;
+        int prev_i = (i - 1) / steps;
+
+        // add up integrals using the trapezoidal rule
+        this->m += (sliced_bits[prev_i] + sliced_bits[curr_i] / 2.0);
+
+        // write data point of audio curve to output-buffer
+        out[i] = cos(2 * M_PI * this->t * fc_sr - 2 * M_PI * m * fd_sr);
+        this->t++;
+    }
```

Comment:
A few things that seem problematic about this DSP:
* It keeps integrating forever in `this->m`. I wonder if errors tend to build up.
* It seems that `out[0]` is never written to. This is bad. Since the algorithm seems to need the previous bit, the block probably needs to use history.
* I'm not sure that `(sliced_bits[prev_i] + sliced_bits[curr_i] / 2.0)` is right. It seems that `(sliced_bits[prev_i] + sliced_bits[curr_i]) / 2.0` was intended instead.
* The phase of the `cos` grows without bounds (due to the `this->t++`). This is bad because evaluation of `cos()` loses accuracy for large phases. The phase should be reduced modulo `2*PI` or similar. In fact, there is no need to have separate variables `this->m` and `this->t`. We can have a single `d_phase` where we add `GR_M_PI * (2.0 * fc_sr - (sliced_bits[prev_i] + sliced_bits[curr_i]))` in each loop iteration, and keep wrapping this `d_phase` modulo `2*PI`.

As an additional remark, the only fundamental difference between this algorithm and the one in [`fxpt_vco.h`](https://github.com/gnuradio/gnuradio/blob/main/gnuradio-runtime/include/gnuradio/fxpt_vco.h) which is what is used by the VCO block is that the phase increment of the VCO is proportional to `(sliced_bits[prev_i] + sliced_bits[curr_i]) / 2.0` instead of simply `sliced_bits[curr_i]`. This only makes a difference at symbol boundaries (and I'm not sure which choice is better; it probably depends on the definition of "better"). In any case, this new block could use `fxpt_vco.h` (which is good, because it's a fixed-point implementation), in the same way that the VCO block does.

Reply:
Thank you for that sharp analysis. I appreciate your thoughts on this and the hint to the fxpt_vco. I will have a look on that block.

For `this->m`, the formular implemented there (and mentioned in the linked blockpost), asks for an integral over the whole signal. The function of `m` is mainly, to "bend" the signal smoothly, when switching the frequencies. As the signal oscillates between 1 and -1, it shouldn't build up errors, as the influence of the first samples decreases with the increasing amount of new samples, right? (But this is more like a gut feeling on my side)

For the history, I partly agree. It would be much more elegant to use history here. I faced the problem, that I can only get the last N input-items. But for aligning the output signals without a break, I need the last N output items, which is not supported by the runtime. So my workaround was to keep the state in the class attributes.

For the `this->t++` I defnitely agree, that this isn't ideal. Your solution with `d_pahse` sounds very interesting. Thank you for pointing that out. I will have a look on that.